### ISSUE TITLE
Avoid double-publishing form updating new dataset

### ISSUE THREAD

**Issue by ktuite** [getodk/central#554]
See notes here:
https://github.com/getodk/central/issues/514#issuecomment-1833712753

When you upload a form that updates a dataset, but the dataset doesn't exist yet, the dataset can't be attached and the draft view of the form doesn't work well. 

You have to publish the form first to get the dataset created, and then you can work with the form draft to attach the dataset to it.

Is this confusing enough that we want to solve it now?

**Comment by lognaturel** [comment]
Could we do something like on form publish, after datasets are created, check if there is an expected attachment that matches a dataset name and isn’t yet populated and if so link to the corresponding dataset?

**Comment by lognaturel** [comment]
I think the pattern from the forum post I linked to will be common — attach the dataset to a registration form to avoid duplicates.

**Comment by ktuite** [comment]
> Could we do something like on form publish, after datasets are created, check if there is an expected attachment that matches a dataset name and isn’t yet populated and if so link to the corresponding dataset?

Maybe! But one related issue (from slack) is how when a form is published and there's no draft, you can't see what the current attachments are. So no one will be able to directly see "oh, this suddenly got attached to this entity list".

**Comment by lognaturel** [comment]
> when a form is published and there's no draft, you can't see what the current attachments are

True! That would be nice to improve at some point. 

> no one will be able to directly see "oh, this suddenly got attached to this entity list".

But at least the implicit behavior would be the expected one!

**Comment by ktuite** [comment]
Looking into how to achieve this is making me want to revive this issue: https://github.com/getodk/central-backend/issues/755

**Comment by dbemke** [comment]
After manually attaching a CSV file in Draft section when the dateset hadn't been created before and publishing the form there isn't the "Link Entity List" button in Draft afterwards.
I used a form that updates entities (Basic updates form which was created just for testing so it's not very meaning)
https://staging.getodk.cloud/#/projects/65/forms/basic_update/draft

**Steps to reproduce:**
Preconditions: In a project the dataset defined in the form doesn’t exist yet
1. In a project upload a draft of a form which updates entities.
2. In Draft – Attachments upload a CSV file.
3. Publish the form.
4. Click Create a new draft and go to Attachments tab.
5. Try to link the dataset.

**Screenshots**
After adding CSV manually and publishing


![Image](https://github.com/getodk/central/assets/56479916/1de594b7-8498-48d3-8e74-900a2bf1320c)

Draft with the button


![Image](https://github.com/getodk/central/assets/56479916/605ede00-35ef-4c3a-b0d5-cbe83d2b9c32)

If I upload a CSV file in a form that was published before and publish it again the issue with the missing button doesn't occur.

**Comment by lognaturel** [comment]
I agree that's a bug! @ktuite want to make sure you saw this one. 😊

**Comment by ktuite** [comment]
I don't think I'm able to reproduce this. 

The only scenario I can get to that doesn't have a "Link to Entity List" button is after step 2, when the form hasn't been published yet. At this point, since the form hasn't been published, neither has the entity list, so there's no button. There's also no pink/white info box saying "Testing Entities..." 

Example: https://staging.getodk.cloud/#/projects/68/forms/participant_create_update/draft/attachments 

After the form has been published, then I seem to always see the "Link" button and info box, like here:
https://staging.getodk.cloud/#/projects/68/forms/basic_update/draft/attachments

My main clue here is that in the screenshot above, when there's no link button, there's also no info box, so it would have to have gotten into a state where the form attachment CSV name didn't match any existing datasets.

**Comment by lognaturel** [comment]
Oh, so interesting! Could it be some kind of frontend caching? You tried publishing with a CSV attachment and immediately starting a new draft?

**Comment by lognaturel** [comment]
>  Could it be some kind of frontend caching? You tried publishing with a CSV attachment and immediately starting a new draft?

I reproduced with the `basic_update` form from above! I immediately created a new draft and that showed the csv attached, same as @dbemke's first screenshot. Then I refreshed and I saw what I expected to see.

I wasn't able to reproduce with `participant_create_update`, weirdly.

**Comment by ktuite** [comment]
Ahh okay, I was finally able to reproduce it. For some reason, the page didn't fetch the list of datasets. Maybe we hadn't refetched the project to realize there were datasets. [These lines of code](https://github.com/getodk/central-frontend/blob/master/src/components/form-attachment/list.vue#L196-L201) are probably the ones to change.

**Comment by matthew-white** [comment]
Interesting! Frontend resets and/or refetches a lot of data after a form draft is published, including the project (in `afterPublish()` in `FormDraftStatus`). It doesn't seem to reset the `datasets` resource, which it probably should. Though it sounds like the issue is that `datasets` isn't even requested in the first place. 🤔 Curious!

**Comment by matthew-white** [comment]
> It doesn't seem to reset the `datasets` resource, which it probably should.

Actually, it looks like `FormShow` does that. `FormDraftStatus` refetches the form, which causes `FormShow` to clear `datasets`. Hard to keep track of some of this!

**Comment by matthew-white** [comment]
Ah never mind, that's `ProjectShow`, not `FormShow`. I'm back to thinking that something probably needs to reset the `datasets` resource after publish (but again, that's probably not the root cause here).

**Comment by srujner** [comment]
@ktuite @matthew-white 
Should we wait for some additional fixes here or will this be filled as separate one? 

If separate, we can close this issue.

**Comment by ktuite** [comment]
I've made a new issue (https://github.com/getodk/central-frontend/issues/916) so this one can be closed!

**Comment by dbemke** [comment]
Tested with success!

**Comment by srujner** [comment]
Tested with success!
